File: tutorials/workflow/go/history-propagation/workflow.go
package main

import (
	"fmt"
	"strings"
	"time"

	"github.com/dapr/durabletask-go/api/protos"
	"github.com/dapr/durabletask-go/workflow"
)

// PatientIntake is the top-level workflow. It verifies the patient's
// insurance, then calls PrescribeMedication as a child workflow with
// PropagateLineage(), giving PrescribeMedication ancestral history to
// forward (or not) downstream.
func PatientIntake(ctx *workflow.WorkflowContext) (any, error) {
	var rec PatientRecord
	if err := ctx.GetInput(&rec); err != nil {
		return nil, err
	}

	if !ctx.IsReplaying() {
		fmt.Printf("  [PatientIntake] Starting intake for patient %s\n", rec.PatientID)
		fmt.Println("  [PatientIntake] Step 1: CallActivity(VerifyInsurance) — no propagation")
	}
	var insured bool
	if err := ctx.CallActivity(VerifyInsurance,
		workflow.WithActivityInput(rec),
	).Await(&insured); err != nil {
		return nil, fmt.Errorf("insurance verification failed: %w", err)
	}
	if !insured {
		return "intake declined: insurance not on file", nil
	}
	if !ctx.IsReplaying() {
		fmt.Println("  [PatientIntake] Step 1 complete: insurance verified")
		fmt.Println("  [PatientIntake] Step 2: CallChildWorkflow(PrescribeMedication)")
		fmt.Println("                  -> WithHistoryPropagation(PropagateLineage)")
	}
	var result string
	if err := ctx.CallChildWorkflow(PrescribeMedication,
		workflow.WithChildWorkflowInput(rec),
		workflow.WithHistoryPropagation(
			workflow.PropagateLineage()),
	).Await(&result); err != nil {
		return nil, fmt.Errorf("prescribing failed: %w", err)
	}
	if !ctx.IsReplaying() {
		fmt.Printf("  [PatientIntake] COMPLETE: %s\n", result)
	}
	return result, nil
}

// VerifyInsurance checks that the patient has active coverage on file.
func VerifyInsurance(ctx workflow.ActivityContext) (any, error) {
	var rec PatientRecord
	if err := ctx.GetInput(&rec); err != nil {
		return false, err
	}
	fmt.Printf("  [VerifyInsurance] Checking coverage for patient %s\n", rec.PatientID)
	return true, nil
}

// PrescribeMedication orchestrates an e-prescription. It checks the patient's
// allergies, screens for drug interactions, runs a compliance audit (as a
// child wf with full lineage), and dispenses the medication (as an activity
// with workflow-level propagation only).
func PrescribeMedication(ctx *workflow.WorkflowContext) (any, error) {
	var rec PatientRecord
	if err := ctx.GetInput(&rec); err != nil {
		return nil, err
	}

	if !ctx.IsReplaying() {
		fmt.Printf("  [PrescribeMedication] Starting prescription: %s %.0fmg for %s\n",
			rec.Medication, rec.Dosage, rec.Condition)
	}

	// Step 1: Allergy check (no propagation — plain activity)
	if !ctx.IsReplaying() {
		fmt.Println("  [PrescribeMedication] Step 1: CallActivity(CheckAllergies) — no propagation")
	}
	var allergyClear bool
	if err := ctx.CallActivity(CheckAllergies,
		workflow.WithActivityInput(rec),
	).Await(&allergyClear); err != nil {
		return nil, fmt.Errorf("allergy check failed: %w", err)
	}
	if !allergyClear {
		return "prescription declined: known allergy", nil
	}
	if !ctx.IsReplaying() {
		fmt.Println("  [PrescribeMedication] Step 1 complete: allergy clear")
	}

	// Step 2: Drug interaction screen (no propagation — plain activity)
	if !ctx.IsReplaying() {
		fmt.Println("  [PrescribeMedication] Step 2: CallActivity(ScreenDrugInteractions) — no propagation")
	}
	var interactionsClear bool
	if err := ctx.CallActivity(ScreenDrugInteractions,
		workflow.WithActivityInput(rec),
	).Await(&interactionsClear); err != nil {
		return nil, fmt.Errorf("drug interaction screen failed: %w", err)
	}
	if !interactionsClear {
		return "prescription declined: drug interaction risk", nil
	}
	if !ctx.IsReplaying() {
		fmt.Println("  [PrescribeMedication] Step 2 complete: no interactions")
	}

	// Step 3: Compliance audit as a child wf.
	// PropagateLineage() — include our events AND any ancestral history we received
	if !ctx.IsReplaying() {
		fmt.Println("  [PrescribeMedication] Step 3: CallChildWorkflow(ComplianceAudit)")
		fmt.Println("                        -> WithHistoryPropagation(PropagateLineage)")
	}
	var compliance ComplianceResult
	if err := ctx.CallChildWorkflow(ComplianceAudit,
		workflow.WithChildWorkflowInput(rec),
		workflow.WithHistoryPropagation(workflow.PropagateLineage()),
	).Await(&compliance); err != nil {
		return nil, fmt.Errorf("compliance audit failed: %w", err)
	}
	if !compliance.Compliant {
		return fmt.Sprintf("prescription blocked: compliance audit failed (risk=%.2f, reason=%s)",
			compliance.RiskScore, compliance.Reason), nil
	}
	if !ctx.IsReplaying() {
		fmt.Printf("  [PrescribeMedication] Step 3 complete: compliance audit passed (risk=%.2f, %d events verified)\n",
			compliance.RiskScore, compliance.EventCount)
	}

	// Step 4: Dispense the medication.
	// In the happy path we attach PropagateOwnHistory() — the pharmacy
	// receives our events only (no ancestral chain) and can verify that the
	// allergy and interaction screens ran. In the negative scenario
	// (rec.ForwardLineage == false) we deliberately omit propagation, so the
	// pharmacy receives no lineage and refuses to dispense.
	dispenseOpts := []workflow.CallActivityOption{workflow.WithActivityInput(rec)}
	if rec.ForwardLineage {
		dispenseOpts = append(dispenseOpts,
			workflow.WithHistoryPropagation(workflow.PropagateOwnHistory()))
	}
	if !ctx.IsReplaying() {
		if rec.ForwardLineage {
			fmt.Println("  [PrescribeMedication] Step 4: CallActivity(DispenseMedication)")
			fmt.Println("                        -> WithHistoryPropagation(PropagateOwnHistory)")
		} else {
			fmt.Println("  [PrescribeMedication] Step 4: CallActivity(DispenseMedication)")
			fmt.Println("                        -> NO history propagation (negative scenario)")
		}
	}
	var dispense DispenseResult
	if err := ctx.CallActivity(DispenseMedication, dispenseOpts...).Await(&dispense); err != nil {
		return nil, fmt.Errorf("dispense failed: %w", err)
	}
	if dispense.Status != "dispensed" {
		if !ctx.IsReplaying() {
			fmt.Printf("  [PrescribeMedication] Step 4 BLOCKED: pharmacy refused to dispense (%s)\n",
				dispense.Reason)
		}
		return fmt.Sprintf("prescription not dispensed: pharmacy refused (%s)", dispense.Reason), nil
	}
	if !ctx.IsReplaying() {
		fmt.Printf("  [PrescribeMedication] Step 4 complete: dispensed (id=%s, %d events verified)\n",
			dispense.DispenseID, dispense.EventCount)
	}

	result := fmt.Sprintf("dispensed: id=%s, patient=%s, drug=%s %.0fmg",
		dispense.DispenseID, rec.PatientID, rec.Medication, rec.Dosage)
	if !ctx.IsReplaying() {
		fmt.Printf("  [PrescribeMedication] COMPLETE: %s\n", result)
	}
	return result, nil
}

// CheckAllergies looks up the patient's allergy list and clears or blocks
// the prescription.
func CheckAllergies(ctx workflow.ActivityContext) (any, error) {
	var rec PatientRecord
	if err := ctx.GetInput(&rec); err != nil {
		return false, err
	}

	ph := ctx.GetPropagatedHistory()
	fmt.Printf("  [CheckAllergies] Screening %s for %s (propagated history: %s)\n",
		rec.PatientID, rec.Medication, describeHistory(ph))
	return true, nil
}

// ScreenDrugInteractions checks the candidate prescription against the
// patient's active medication list.
func ScreenDrugInteractions(ctx workflow.ActivityContext) (any, error) {
	var rec PatientRecord
	if err := ctx.GetInput(&rec); err != nil {
		return false, err
	}

	ph := ctx.GetPropagatedHistory()
	fmt.Printf("  [ScreenDrugInteractions] Screening %s %.0fmg for %s (propagated history: %s)\n",
		rec.Medication, rec.Dosage, rec.PatientID, describeHistory(ph))
	return true, nil
}

// ComplianceAudit is a child wf that inspects the parent's propagated
// history to verify the prescribing pipeline ran correctly. It refuses
// to approve dispensing unless the required upstream steps
// (insurance, allergies, interactions) are all present and completed.
func ComplianceAudit(ctx *workflow.WorkflowContext) (any, error) {
	var rec PatientRecord
	if err := ctx.GetInput(&rec); err != nil {
		return nil, err
	}

	if !ctx.IsReplaying() {
		fmt.Printf("  [ComplianceAudit] Auditing prescription for patient %s\n", rec.PatientID)
	}

	history := ctx.GetPropagatedHistory()
	if history == nil {
		if !ctx.IsReplaying() {
			fmt.Println("  [ComplianceAudit] WARNING: No propagated history received!")
			fmt.Println("  [ComplianceAudit] BLOCKED — cannot verify upstream pipeline without history")
		}
		return ComplianceResult{
			Compliant: false,
			RiskScore: 1.0,
			Reason:    "no execution history provided — cannot verify caller pipeline",
		}, nil
	}

	if !ctx.IsReplaying() {
		fmt.Printf("  [ComplianceAudit] Received propagated history: %d events (scope: %s)\n",
			len(history.Events()), describeScope(history.Scope()))
		fmt.Printf("  [ComplianceAudit] Apps in chain: %v\n", history.GetAppIDs())
		for _, wf := range history.GetWorkflows() {
			fmt.Printf("  [ComplianceAudit]   workflow: app=%s, name=%s, instance=%s\n",
				wf.AppID, wf.Name, wf.InstanceID)
		}
	}

	intakeWf, err := history.GetLastWorkflowByName("PatientIntake")
	if err != nil {
		return ComplianceResult{}, fmt.Errorf("expected PatientIntake in propagated history: %w", err)
	}
	prescribeWf, err := history.GetLastWorkflowByName("PrescribeMedication")
	if err != nil {
		return ComplianceResult{}, fmt.Errorf("expected PrescribeMedication in propagated history: %w", err)
	}
	insurance, err := intakeWf.GetLastActivityByName("VerifyInsurance")
	if err != nil {
		return ComplianceResult{}, fmt.Errorf("expected VerifyInsurance in propagated history: %w", err)
	}
	allergies, err := prescribeWf.GetLastActivityByName("CheckAllergies")
	if err != nil {
		return ComplianceResult{}, fmt.Errorf("expected CheckAllergies in propagated history: %w", err)
	}
	interactions, err := prescribeWf.GetLastActivityByName("ScreenDrugInteractions")
	if err != nil {
		return ComplianceResult{}, fmt.Errorf("expected ScreenDrugInteractions in propagated history: %w", err)
	}

	if !ctx.IsReplaying() {
		fmt.Printf("  [ComplianceAudit] Verification:\n")
		fmt.Printf("  [ComplianceAudit]   PatientIntake/VerifyInsurance: started=%v, completed=%v\n",
			insurance.Started, insurance.Completed)
		fmt.Printf("  [ComplianceAudit]   PrescribeMedication/CheckAllergies: started=%v, completed=%v\n",
			allergies.Started, allergies.Completed)
		fmt.Printf("  [ComplianceAudit]   PrescribeMedication/ScreenDrugInteractions: started=%v, completed=%v\n",
			interactions.Started, interactions.Completed)
	}

	if !insurance.Completed || !allergies.Completed || !interactions.Completed {
		if !ctx.IsReplaying() {
			fmt.Println("  [ComplianceAudit] BLOCKED — required upstream checks not completed")
		}
		return ComplianceResult{
			Compliant:  false,
			RiskScore:  0.9,
			Reason:     "required upstream checks not completed in propagated history",
			EventCount: len(history.Events()),
		}, nil
	}

	riskScore := 0.1
	if rec.Dosage > 1000 {
		riskScore = 0.3
	}

	if !ctx.IsReplaying() {
		fmt.Printf("  [ComplianceAudit] APPROVED (risk=%.2f)\n", riskScore)
	}
	return ComplianceResult{
		Compliant:  true,
		RiskScore:  riskScore,
		Reason:     "all upstream checks verified in propagated history",
		EventCount: len(history.Events()),
	}, nil
}

// DispenseMedication issues the final prescription. The pharmacy refuses to
// dispense unless the propagated workflow history proves the prescribing
// pipeline ran the required allergy and drug-interaction screens. If no
// history is propagated (or it is missing those checks), it returns a
// "refused" result instead of dispensing.
func DispenseMedication(ctx workflow.ActivityContext) (any, error) {
	var rec PatientRecord
	if err := ctx.GetInput(&rec); err != nil {
		return nil, err
	}

	ph := ctx.GetPropagatedHistory()
	fmt.Printf("  [DispenseMedication] Dispense request: %s %.0fmg for %s (propagated history: %s)\n",
		rec.Medication, rec.Dosage, rec.PatientID, describeHistory(ph))

	// Pharmacy policy: no lineage, no dispense. Without propagated history the
	// pharmacy cannot prove the prescription was screened, so it refuses.
	if ph == nil {
		fmt.Printf("  [DispenseMedication] REFUSED — no propagated history; cannot verify screening for %s\n",
			rec.PatientID)
		return DispenseResult{
			Status: "refused",
			Reason: "missing lineage: no propagated history received from prescriber",
		}, nil
	}

	eventCount := len(ph.Events())
	fmt.Printf("  [DispenseMedication] Apps in chain: %v\n", ph.GetAppIDs())
	for _, wf := range ph.GetWorkflows() {
		fmt.Printf("  [DispenseMedication]   workflow: app=%s, name=%s, instance=%s\n",
			wf.AppID, wf.Name, wf.InstanceID)
	}
	scheduledNames := make(map[string]string) // key=taskExecutionId, val=activity name
	for i, event := range ph.Events() {
		if ts := event.GetTaskScheduled(); ts != nil {
			scheduledNames[ts.GetTaskExecutionId()] = ts.GetName()
		}
		fmt.Printf("  [DispenseMedication]   event[%d]: %s\n", i, describeEventResolved(event, scheduledNames))
	}

	// Verify the prescriber's own history shows both screens completed.
	prescribeWf, err := ph.GetLastWorkflowByName("PrescribeMedication")
	if err != nil {
		fmt.Printf("  [DispenseMedication] REFUSED — propagated history is missing the PrescribeMedication lineage for %s\n",
			rec.PatientID)
		return DispenseResult{
			Status:     "refused",
			Reason:     "missing lineage: PrescribeMedication not present in propagated history",
			EventCount: eventCount,
		}, nil
	}
	allergies, errAllergy := prescribeWf.GetLastActivityByName("CheckAllergies")
	interactions, errInteraction := prescribeWf.GetLastActivityByName("ScreenDrugInteractions")
	if errAllergy != nil || errInteraction != nil || !allergies.Completed || !interactions.Completed {
		fmt.Printf("  [DispenseMedication] REFUSED — required screening not verified in propagated history for %s\n",
			rec.PatientID)
		return DispenseResult{
			Status:     "refused",
			Reason:     "missing lineage: allergy/interaction screening not verified in propagated history",
			EventCount: eventCount,
		}, nil
	}

	dispenseID := fmt.Sprintf("rx-%s-%d", rec.PatientID, time.Now().UnixMilli())
	fmt.Printf("  [DispenseMedication] DISPENSED: %s\n", dispenseID)

	return DispenseResult{
		DispenseID: dispenseID,
		Status:     "dispensed",
		EventCount: eventCount,
	}, nil
}

// describeEventResolved returns a human-readable description of a history event.
func describeEventResolved(event *protos.HistoryEvent, scheduledNames map[string]string) string {
	eventType := fmt.Sprintf("%T", event.EventType)
	if idx := strings.LastIndex(eventType, "."); idx >= 0 {
		eventType = eventType[idx+1:]
	}
	switch {
	case event.GetTaskScheduled() != nil:
		return fmt.Sprintf("%s -> %s", eventType, event.GetTaskScheduled().GetName())
	case event.GetTaskCompleted() != nil:
		if name, ok := scheduledNames[event.GetTaskCompleted().GetTaskExecutionId()]; ok {
			return fmt.Sprintf("%s -> %s", eventType, name)
		}
		return eventType
	case event.GetExecutionStarted() != nil:
		return fmt.Sprintf("%s -> %s", eventType, event.GetExecutionStarted().GetName())
	case event.GetChildWorkflowInstanceCreated() != nil:
		return fmt.Sprintf("%s -> %s", eventType, event.GetChildWorkflowInstanceCreated().GetName())
	default:
		return eventType
	}
}

func describeHistory(ph *workflow.PropagatedHistory) string {
	if ph == nil {
		return "none"
	}
	return fmt.Sprintf("%d events, scope=%s", len(ph.Events()), describeScope(ph.Scope()))
}

func describeScope(scope fmt.Stringer) string {
	s := scope.String()
	s = strings.TrimPrefix(s, "HISTORY_PROPAGATION_SCOPE_")
	return s
}

File: tutorials/workflow/python/history-propagation/workflow.py
# -*- coding: utf-8 -*-
# Copyright 2026 The Dapr Authors
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#     http://www.apache.org/licenses/LICENSE-2.0
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Workflow + activity definitions for the history propagation quickstart.

PatientIntake (root workflow)
  |- VerifyInsurance      (activity, no propagation)
  `- PrescribeMedication  (child workflow, PropagationScope.LINEAGE)
        |- CheckAllergies         (activity, no propagation)
        |- ScreenDrugInteractions (activity, no propagation)
        |- ComplianceAudit        (grandchild wf, PropagationScope.LINEAGE)
        |     reads: PatientIntake/VerifyInsurance
        |            PrescribeMedication/CheckAllergies
        |            PrescribeMedication/ScreenDrugInteractions
        `- DispenseMedication     (activity, PropagationScope.OWN_HISTORY
                                   when forward_lineage=True, else no propagation)
              reads: PrescribeMedication events only (no PatientIntake chain)
"""

from __future__ import annotations

import time

import dapr.ext.workflow as wf

from models import ComplianceResult, DispenseResult, PatientRecord

wfr = wf.WorkflowRuntime()


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _describe_scope(scope: wf.PropagationScope | None) -> str:
    if scope is None:
        return 'NONE'
    return scope.name


def _describe_history(history: wf.PropagatedHistory | None) -> str:
    if history is None:
        return 'none'
    return f'{len(history.events)} events, scope={_describe_scope(history.scope)}'


# ---------------------------------------------------------------------------
# Activities
# ---------------------------------------------------------------------------


@wfr.activity(name='VerifyInsurance')
def verify_insurance(ctx: wf.WorkflowActivityContext, rec_json: str) -> bool:
    rec = PatientRecord.from_json(rec_json)
    print(f'  [VerifyInsurance] Checking coverage for patient {rec.patient_id}', flush=True)
    return True


@wfr.activity(name='CheckAllergies')
def check_allergies(ctx: wf.WorkflowActivityContext, rec_json: str) -> bool:
    rec = PatientRecord.from_json(rec_json)
    history = ctx.get_propagated_history()
    print(
        f'  [CheckAllergies] Screening {rec.patient_id} for {rec.medication} '
        f'(propagated history: {_describe_history(history)})',
        flush=True,
    )
    return True


@wfr.activity(name='ScreenDrugInteractions')
def screen_drug_interactions(ctx: wf.WorkflowActivityContext, rec_json: str) -> bool:
    rec = PatientRecord.from_json(rec_json)
    history = ctx.get_propagated_history()
    print(
        f'  [ScreenDrugInteractions] Screening {rec.medication} {rec.dosage:.0f}mg '
        f'for {rec.patient_id} (propagated history: {_describe_history(history)})',
        flush=True,
    )
    return True


@wfr.activity(name='DispenseMedication')
def dispense_medication(ctx: wf.WorkflowActivityContext, rec_json: str) -> str:
    """Refuses to dispense unless the propagated history proves screening ran."""
    rec = PatientRecord.from_json(rec_json)
    history = ctx.get_propagated_history()
    print(
        f'  [DispenseMedication] Dispense request: {rec.medication} {rec.dosage:.0f}mg '
        f'for {rec.patient_id} (propagated history: {_describe_history(history)})',
        flush=True,
    )

    if history is None:
        print(
            f'  [DispenseMedication] REFUSED — no propagated history; '
            f'cannot verify screening for {rec.patient_id}',
            flush=True,
        )
        refused = DispenseResult(
            dispense_id='',
            status='refused',
            reason='missing lineage: no propagated history received from prescriber',
            event_count=0,
        )
        return refused.to_json()

    event_count = len(history.events)
    print(f'  [DispenseMedication] Apps in chain: {history.get_app_ids()}', flush=True)
    for wf_result in history.get_workflows():
        print(
            f'  [DispenseMedication]   workflow: app={wf_result.app_id}, '
            f'name={wf_result.name}, instance={wf_result.instance_id}',
            flush=True,
        )

    try:
        prescribe_wf = history.get_last_workflow_by_name('PrescribeMedication')
    except wf.PropagationNotFoundError:
        print(
            f'  [DispenseMedication] REFUSED — propagated history is missing the '
            f'PrescribeMedication lineage for {rec.patient_id}',
            flush=True,
        )
        refused = DispenseResult(
            dispense_id='',
            status='refused',
            reason='missing lineage: PrescribeMedication not present in propagated history',
            event_count=event_count,
        )
        return refused.to_json()

    try:
        allergies_act = prescribe_wf.get_last_activity_by_name('CheckAllergies')
        interactions_act = prescribe_wf.get_last_activity_by_name('ScreenDrugInteractions')
    except wf.PropagationNotFoundError:
        print(
            f'  [DispenseMedication] REFUSED — required screening not verified in '
            f'propagated history for {rec.patient_id}',
            flush=True,
        )
        refused = DispenseResult(
            dispense_id='',
            status='refused',
            reason='missing lineage: allergy/interaction screening not verified in propagated history',
            event_count=event_count,
        )
        return refused.to_json()

    if not (allergies_act.completed and interactions_act.completed):
        print(
            f'  [DispenseMedication] REFUSED — required screening not verified in '
            f'propagated history for {rec.patient_id}',
            flush=True,
        )
        refused = DispenseResult(
            dispense_id='',
            status='refused',
            reason='missing lineage: allergy/interaction screening not verified in propagated history',
            event_count=event_count,
        )
        return refused.to_json()

    dispense_id = f'rx-{rec.patient_id}-{int(time.time() * 1000)}'
    print(f'  [DispenseMedication] DISPENSED: {dispense_id}', flush=True)
    dispensed = DispenseResult(
        dispense_id=dispense_id,
        status='dispensed',
        reason='',
        event_count=event_count,
    )
    return dispensed.to_json()


# ---------------------------------------------------------------------------
# Child workflows
# ---------------------------------------------------------------------------


@wfr.workflow(name='ComplianceAudit')
def compliance_audit(ctx: wf.DaprWorkflowContext, rec_json: str):
    """Approves only if the propagated history shows insurance + screening completed."""
    rec = PatientRecord.from_json(rec_json)
    if not ctx.is_replaying:
        print(
            f'  [ComplianceAudit] Auditing prescription for patient {rec.patient_id}',
            flush=True,
        )

    history = ctx.get_propagated_history()
    if history is None:
        if not ctx.is_replaying:
            print('  [ComplianceAudit] WARNING: No propagated history received!', flush=True)
            print(
                '  [ComplianceAudit] BLOCKED — cannot verify upstream pipeline without history',
                flush=True,
            )
        no_history = ComplianceResult(
            compliant=False,
            risk_score=1.0,
            reason='no execution history provided — cannot verify caller pipeline',
            event_count=0,
        )
        return no_history.to_json()

    if not ctx.is_replaying:
        print(
            f'  [ComplianceAudit] Received propagated history: '
            f'{len(history.events)} events (scope: {_describe_scope(history.scope)})',
            flush=True,
        )
        print(f'  [ComplianceAudit] Apps in chain: {history.get_app_ids()}', flush=True)
        for wf_result in history.get_workflows():
            print(
                f'  [ComplianceAudit]   workflow: app={wf_result.app_id}, '
                f'name={wf_result.name}, instance={wf_result.instance_id}',
                flush=True,
            )

    try:
        intake_wf = history.get_last_workflow_by_name('PatientIntake')
        prescribe_wf = history.get_last_workflow_by_name('PrescribeMedication')
        insurance = intake_wf.get_last_activity_by_name('VerifyInsurance')
        allergies = prescribe_wf.get_last_activity_by_name('CheckAllergies')
        interactions = prescribe_wf.get_last_activity_by_name('ScreenDrugInteractions')
    except wf.PropagationNotFoundError as exc:
        if not ctx.is_replaying:
            print(f'  [ComplianceAudit] BLOCKED — {exc}', flush=True)
        missing = ComplianceResult(
            compliant=False,
            risk_score=0.9,
            reason=f'required item missing from propagated history: {exc}',
            event_count=len(history.events),
        )
        return missing.to_json()

    if not ctx.is_replaying:
        print('  [ComplianceAudit] Verification:', flush=True)
        print(
            f'  [ComplianceAudit]   PatientIntake/VerifyInsurance: '
            f'completed={insurance.completed}',
            flush=True,
        )
        print(
            f'  [ComplianceAudit]   PrescribeMedication/CheckAllergies: '
            f'completed={allergies.completed}',
            flush=True,
        )
        print(
            f'  [ComplianceAudit]   PrescribeMedication/ScreenDrugInteractions: '
            f'completed={interactions.completed}',
            flush=True,
        )

    all_completed = insurance.completed and allergies.completed and interactions.completed
    if not all_completed:
        if not ctx.is_replaying:
            print(
                '  [ComplianceAudit] BLOCKED — required upstream checks not completed',
                flush=True,
            )
        blocked = ComplianceResult(
            compliant=False,
            risk_score=0.9,
            reason='required upstream checks not completed in propagated history',
            event_count=len(history.events),
        )
        return blocked.to_json()

    risk_score = 0.3 if rec.dosage > 1000 else 0.1
    if not ctx.is_replaying:
        print(f'  [ComplianceAudit] APPROVED (risk={risk_score:.2f})', flush=True)
    approved = ComplianceResult(
        compliant=True,
        risk_score=risk_score,
        reason='all upstream checks verified in propagated history',
        event_count=len(history.events),
    )
    return approved.to_json()


@wfr.workflow(name='PrescribeMedication')
def prescribe_medication(ctx: wf.DaprWorkflowContext, rec_json: str):
    rec = PatientRecord.from_json(rec_json)
    if not ctx.is_replaying:
        print(
            f'  [PrescribeMedication] Starting prescription: {rec.medication} '
            f'{rec.dosage:.0f}mg for {rec.condition}',
            flush=True,
        )

    if not ctx.is_replaying:
        print(
            '  [PrescribeMedication] Step 1: CallActivity(CheckAllergies) — no propagation',
            flush=True,
        )
    allergy_clear = yield ctx.call_activity(check_allergies, input=rec_json)
    if not allergy_clear:
        return 'prescription declined: known allergy'
    if not ctx.is_replaying:
        print('  [PrescribeMedication] Step 1 complete: allergy clear', flush=True)

    if not ctx.is_replaying:
        print(
            '  [PrescribeMedication] Step 2: CallActivity(ScreenDrugInteractions) '
            '— no propagation',
            flush=True,
        )
    interactions_clear = yield ctx.call_activity(screen_drug_interactions, input=rec_json)
    if not interactions_clear:
        return 'prescription declined: drug interaction risk'
    if not ctx.is_replaying:
        print('  [PrescribeMedication] Step 2 complete: no interactions', flush=True)

    if not ctx.is_replaying:
        print(
            '  [PrescribeMedication] Step 3: CallChildWorkflow(ComplianceAudit)',
            flush=True,
        )
        print(
            '                        -> propagation=PropagationScope.LINEAGE',
            flush=True,
        )
    audit_json = yield ctx.call_child_workflow(
        compliance_audit,
        input=rec_json,
        propagation=wf.PropagationScope.LINEAGE,
    )
    audit = ComplianceResult.from_json(audit_json)
    if not audit.compliant:
        return (
            f'prescription blocked: compliance audit failed '
            f'(risk={audit.risk_score:.2f}, reason={audit.reason})'
        )
    if not ctx.is_replaying:
        print(
            f'  [PrescribeMedication] Step 3 complete: compliance audit passed '
            f'(risk={audit.risk_score:.2f}, {audit.event_count} events verified)',
            flush=True,
        )

    # Step 4 demonstrates the two propagation modes side-by-side: forward_lineage=True
    # attaches OWN_HISTORY so the pharmacy can verify upstream screening;
    # forward_lineage=False omits propagation, so the pharmacy refuses to dispense.
    if not ctx.is_replaying:
        print(
            '  [PrescribeMedication] Step 4: CallActivity(DispenseMedication)',
            flush=True,
        )
        if rec.forward_lineage:
            print(
                '                        -> propagation=PropagationScope.OWN_HISTORY',
                flush=True,
            )
        else:
            print(
                '                        -> NO history propagation (negative scenario)',
                flush=True,
            )
    if rec.forward_lineage:
        dispense_json = yield ctx.call_activity(
            dispense_medication,
            input=rec_json,
            propagation=wf.PropagationScope.OWN_HISTORY,
        )
    else:
        dispense_json = yield ctx.call_activity(dispense_medication, input=rec_json)
    dispense = DispenseResult.from_json(dispense_json)
    if dispense.status != 'dispensed':
        if not ctx.is_replaying:
            print(
                f'  [PrescribeMedication] Step 4 BLOCKED: pharmacy refused to dispense '
                f'({dispense.reason})',
                flush=True,
            )
        return f'prescription not dispensed: pharmacy refused ({dispense.reason})'
    if not ctx.is_replaying:
        print(
            f'  [PrescribeMedication] Step 4 complete: dispensed '
            f'(id={dispense.dispense_id}, {dispense.event_count} events verified)',
            flush=True,
        )

    result = (
        f'dispensed: id={dispense.dispense_id}, patient={rec.patient_id}, '
        f'drug={rec.medication} {rec.dosage:.0f}mg'
    )
    if not ctx.is_replaying:
        print(f'  [PrescribeMedication] COMPLETE: {result}', flush=True)
    return result


@wfr.workflow(name='PatientIntake')
def patient_intake(ctx: wf.DaprWorkflowContext, rec_json: str):
    rec = PatientRecord.from_json(rec_json)
    if not ctx.is_replaying:
        print(f'  [PatientIntake] Starting intake for patient {rec.patient_id}', flush=True)
        print(
            '  [PatientIntake] Step 1: CallActivity(VerifyInsurance) — no propagation',
            flush=True,
        )
    insured = yield ctx.call_activity(verify_insurance, input=rec_json)
    if not insured:
        return 'intake declined: insurance not on file'
    if not ctx.is_replaying:
        print('  [PatientIntake] Step 1 complete: insurance verified', flush=True)
        print('  [PatientIntake] Step 2: CallChildWorkflow(PrescribeMedication)', flush=True)
        print(
            '                  -> propagation=PropagationScope.LINEAGE',
            flush=True,
        )
    result = yield ctx.call_child_workflow(
        prescribe_medication,
        input=rec_json,
        propagation=wf.PropagationScope.LINEAGE,
    )
    if not ctx.is_replaying:
        print(f'  [PatientIntake] COMPLETE: {result}', flush=True)
    return result
